pkg/spdk/engine.go

```diff
@@ -1229,6 +1304,388 @@ func (e *Engine) validateAndUpdateReplicaNvme(replicaName string, bdev *spdktype
 	return e.ReplicaStatusMap[replicaName].Mode, nil
 }
 
+// This method performs an online volume expansion for the Longhorn Engine using SPDK. It:
+// Expands underlying replica logical volumes (lvol)
+// Recreates the SPDK RAID bdev
+// Suspends and resumes frontend I/O as needed
+// Ensures cleanup and status update on failure
+func (e *Engine) Expand(spdkClient *spdkclient.Client, size uint64) (err error) {
+	e.Lock()
+	defer e.Unlock()
+
+	e.log.WithFields(logrus.Fields{
+		"engineName": e.Name,
+		"volumeName": e.VolumeName,
+		"frontend":   e.Frontend,
+		"specSize":   e.SpecSize,
+		"newSize":    size,
+	}).Info("Expanding engine")
+
+	if err := e.startExpansion(size); err != nil {
+		return errors.Wrap(err, "startExpansion failed")
+	}
+
+	var (
+		expanded  = false
+		engineErr error
+	)
+
+	// engineErr will be set when the engine failed to do any non-recoverable operations.
+	defer func() {
+		if err != nil {
+			e.log.WithError(engineErr).Errorf("Engine %s failed to expand", e.Name)
+			e.lastExpansionError = errors.Wrap(err, "engine failed to expand expansion").Error()
+			e.lastExpansionFailedAt = time.Now().UTC().Format(time.RFC3339Nano)
+		}
+
+		if engineErr != nil {
+			if e.State != types.InstanceStateError {
+				e.State = types.InstanceStateError
+			}
+			e.ErrorMsg = engineErr.Error()
+
+			if errUpdateLogger := e.log.UpdateLogger(logrus.Fields{
+				"replicaStatusMap": e.ReplicaStatusMap,
+			}); errUpdateLogger != nil {
+				e.log.WithError(errUpdateLogger).Warn("Failed to update logger with replica status map during engine creation")
+			}
+
+			e.lastExpansionError = errors.Wrap(err, "engine under non-recoverable operation").Error()
+			e.lastExpansionFailedAt = time.Now().UTC().Format(time.RFC3339Nano)
+		}
+
+		e.finishExpansion(expanded, size)
+	}()
+
+	// fetch all replica clients
+	replicaClients, err := e.getReplicaClients()
+	if err != nil {
+		return err
+	}
+	defer e.closeRplicaClients(replicaClients)
+	if len(replicaClients) == 0 || e.isAnyReplicaUnderRebuilding(replicaClients) {
+		return errors.New("expansion failed: no healthy replica or at least one replica is under rebuilding")
+	}
+
+	// prepareRaidForExpansion checks if RAID exists, suspends frontend if needed, and deletes the RAID bdev.
+	isSusupend, bdevRaidUUID, err := e.prepareRaidForExpansion(spdkClient)
```

Comment:
The returned error of this func is probably an engine error, which is typically related to the frontend or RAID and probably affects the IO. For this kind of error, we need to crash the engine.